BT Classic: support for 32 and 128-bit custom UUIDs (IDFGH-10721)

ilutchenko opened the pull request:
## The problem

This PR fixes issue #11529  

## What was done

I added new API functions that can add 32 and 128-bit UUID to the EIR data when these UUIDs are set in SDP.
The old functions that only work with 16-bit UUIDs have been left unchanged to avoid having to redo code that already utilizes them.

## How it was tested

`sdptool` and `wireshark` output:
for 32-bit UUID:
![image](https://github.com/espressif/esp-idf/assets/6449804/7e4bb1d5-3574-412d-ac2a-c2a7118b842a)
![image](https://github.com/espressif/esp-idf/assets/6449804/b7097b32-9108-44f7-983b-dce2efcdb4f9)

for 128-bit UUID:
![image](https://github.com/espressif/esp-idf/assets/6449804/beb1db69-3105-4d75-89de-775d2e1025b3)
![image](https://github.com/espressif/esp-idf/assets/6449804/9d7a7548-a72b-4ec9-8cdd-493e1b4b5c0c)

xiongweichao commented on components/bt/host/bluedroid/stack/btm/btm_inq.c:
```diff
@@ -2537,7 +2537,69 @@ void BTM_AddEirService( UINT32 *p_eir_uuid, UINT16 uuid16 )
+                *empty_slot = xx;
+        } else if ( *match_slot == BTA_EIR_SERVER_NUM_CUSTOM_UUID ) {
+            if ( !memcmp (&custom_uuid[xx].uu, &uuid.uu, uuid.len) ) {
+                *match_slot = xx;
```
If it has been matched, can it be returned directly?

ilutchenko replied:
Yes.
I split this search algorithm into `btm_find_empty_custom_uuid_slot()` and `btm_find_match_custom_uuid_slot()`.
This should make the code simpler.

xiongweichao commented on components/bt/host/bluedroid/stack/btm/btm_inq.c:
```diff
@@ -2537,7 +2537,69 @@ void BTM_AddEirService( UINT32 *p_eir_uuid, UINT16 uuid16 )
+            return;
+        } else {
+            memcpy (&(custom_uuid[empty_slot]), &(uuid), sizeof(tBT_UUID));
+            BTM_TRACE_EVENT("UUID saved in %d slot\r\n", empty_slot);
```
delete `\r\n`

xiongweichao commented on components/bt/host/bluedroid/stack/btm/btm_inq.c:
```diff
@@ -2537,7 +2537,69 @@ void BTM_AddEirService( UINT32 *p_eir_uuid, UINT16 uuid16 )
+            BTM_TRACE_EVENT("UUID saved in %d slot\r\n", empty_slot);
+        }
+    } else {
+        BTM_TRACE_WARNING("UUID is already added for EIR");
```
If the UUID has been added to the EIR, there should be no need to call `bta_dm_set_eir` in `bta_dm_eir_update_uuid`.

ilutchenko replied:
I added the `BTM_HasCustomEirService()` function to avoid this situation.

xiongweichao commented on components/bt/host/bluedroid/btc/profile/std/sdp/btc_sdp.c:
```diff
@@ -331,7 +327,13 @@ static int add_raw_sdp(const bluetooth_sdp_record* rec)
     } else {
-        bta_sys_add_uuid(service);
+        if (rec->hdr.bt_uuid.len == ESP_UUID_LEN_16){
+            bta_sys_add_uuid(rec->hdr.bt_uuid.uuid.uuid16);
```
The uuid should also be removed when [deleting](https://github.com/espressif/esp-idf/blob/28167ea5a3cdd49227216c4131a9df7f4eda260e/components/bt/host/bluedroid/btc/profile/std/sdp/btc_sdp.c#L828) the sdp record.

xiongweichao commented:
Please delete unnecessary whitespace in the code.

ESP-YTGerd commented:
Hi, @ilutchenko thanks for your contribution, are you still following up this PR? Please  resolve all the comments so that we can merge it.

ilutchenko commented:
Hi @ESP-YTGerd ! Yes, I will fix up it soon. Just found some bugs during applying @xiongweichao's suggestions, need to figure out how to deal with them in the best way.

ilutchenko commented:
So, deleting records and EIR UUIDs pulled me into a deep rabbit hole.
As I discovered, SDP records creation and removal didn't work correctly.
So I spent some time debugging and bugfixing.

Done since the previous PR version:
- resolved **xiongweichao's** comments
- fixed handler return when SDP record is created
- fixed SDP database slot release in `free_sdp_slot()`
- UUID is removed from EIR when you delete the SDP record with `esp_sdp_remove_record()`

@xiongweichao , could you take another look at this PR, please?

xiongweichao commented on components/bt/host/bluedroid/stack/btm/btm_inq.c:
```diff
@@ -2537,7 +2537,101 @@ void BTM_AddEirService( UINT32 *p_eir_uuid, UINT16 uuid16 )
+{
+    for (UINT8 xx = 0; xx < BTA_EIR_SERVER_NUM_CUSTOM_UUID; xx++) {
+        if (!memcmp (&custom_uuid[xx].uu, &uuid.uu, uuid.len)) {
+            return xx;
```
Maybe there will be a mismatch between `custom_uuid[xx]`(`custom_uuid.len = LEN_UUID_128`) and `uuid`(`uuid.len = LEN_UUID_16`)? What do you think about matching tBT_UUID?

ilutchenko replied:
Great catch!
I added a small `btm_compare_uuid()` that compares only UUIDs with similar lengths.

xiongweichao commented on components/bt/host/bluedroid/bta/dm/bta_dm_act.c:
```diff
@@ -4433,21 +4433,38 @@ static void bta_dm_eir_search_services( tBTM_INQ_RESULTS  *p_result,
-        return;
-    }
+    /* 32 and 128-bit UUIDs go to the bta_dm_cb.custom_uuid array */
+    if ((uuid.len == LEN_UUID_32) || (uuid.len == LEN_UUID_128)){
```
`)` and `{` add a space between. Other places also need to be modified.

xiongweichao commented on components/bt/host/bluedroid/btc/profile/std/sdp/btc_sdp.c:
```diff
@@ -248,7 +286,7 @@ static int free_sdp_slot(int id)
     }
     osi_free(slot);
-    slot = NULL;
+    sdp_local_param.sdp_slots[id] = NULL;
```
A mutex is required.

ilutchenko replied:
Hmm, under this line I moved the mutex release from several lines above.
Seems to me that it doesn't break anything.

ilutchenko commented:
@xiongweichao , @ESP-YTGerd 
Could you take a last look at this PR, please?

BTW, I have a question about failed CI. Pre-commit check passed successfully on my local machine but failed on GitHub Actions with a strange error.
Should I do something with it?

ESP-YTGerd commented:
Hi, @ilutchenko
1: Can try to rebase to the `Origin/Master` branch, since the pre-commit just check the codes what you have committed.
2: It's also recommended to modify the commit message, which has the following format:
 ```
Subject (first line) has format: <type|action>(<scope/component>): <summary> (or <type|action>: <summary>)
If related to Github issue, add **full** Github issue link, e.g. `https://github.com/espressif/esp-idf/issues/47`
 ```  
Example of a complete commit message:
```
feat(bt/bluedroid): Add new APIs for 32 and 128-bit UUIDs

1: <detail >(if needed)

Closes https://github.com/espressif/esp-idf/issues/xxx (if github issue linked)
```
Explanation of typical `<action/type>` in conventional commit style:
```
* `change`: making general changes or improvements to the codebase (e.g.: `change(api): Style of authentication module`)
* `ci`: making changes to the CI system configuration or CI scripts (e.g.: `ci: Configure pre-commit checks`)
* `docs`: making changes to the documentation (e.g.: `docs(ble): Update BLE API documentation`)
* `feat`: adding a new feature or functionality to the project (e.g: `feat(api): Implement user login functionality`)
* `fix`: fixing a bug or resolving an issue (e.g: `fix(mbedtls): Fix null pointer exception in data validation`)
* `refactor`: refactoring existing code without changing its behavior (e.g: `refactor(area): Extract common utility function`)
* `remove`: removing files, features, or functionalities from the project (e.g: `remove(api): Delete deprecated API endpoint`)
* `revert`: reverting a previous change (e.g: `revert(api): Revert changes to user authentication from MR 12345`)
```

ilutchenko commented:
Hi @ESP-YTGerd 
So, I squashed all the work into one commit and updated its description as you suggested.
I think it is ready to merge.

ESP-YTGerd commented:
Hi, @ilutchenko
LGTM, but also please resolve the comments what I have reviewed and then I can merge this PR, thanks!

ilutchenko commented:
Hi @ESP-YTGerd ,
All notes were implemented and comments sections were resolved.

ESP-YTGerd commented:
@ilutchenko 
oh, sorry, my bad, I forgot to  submit my comments.

ilutchenko commented on components/bt/host/bluedroid/stack/include/stack/btm_api.h:
```diff
@@ -4095,6 +4112,21 @@ tBTM_EIR_SEARCH_RESULT BTM_HasInquiryEirService( tBTM_INQ_RESULTS *p_results,
+**
+*******************************************************************************/
+//extern
+void BTM_AddCustomEirService( tBT_UUID *custom_uuid, tBT_UUID uuid );
```
Hmm, some other function declarations in this file have spaces near bracers, and some don't.
Okay, I removed them.

ilutchenko commented:
@ESP-YTGerd , done

xiongweichao commented on components/bt/host/bluedroid/stack/btm/btm_inq.c:
```diff
@@ -2535,9 +2535,125 @@ void BTM_AddEirService( UINT32 *p_eir_uuid, UINT16 uuid16 )
+    UINT8 empty_slot = btm_find_empty_custom_uuid_slot(custom_uuid, uuid);
+
+    if (empty_slot == BTA_EIR_SERVER_NUM_CUSTOM_UUID) {
+        BTM_TRACE_ERROR("No space to add UUID for EIR");
```
`BTM_TRACE_ERROR` ->  `BTM_TRACE_WARNING`

xiongweichao commented on components/bt/host/bluedroid/stack/include/stack/btm_api.h:
```diff
@@ -4080,6 +4080,23 @@ BOOLEAN BTM_HasEirService( UINT32 *p_eir_uuid, UINT16 uuid16 );
+**                  FALSE - if not found
+**
+*******************************************************************************/
+//extern
```
Delete it.

xiongweichao commented on components/bt/host/bluedroid/stack/include/stack/btm_api.h:
```diff
@@ -4095,6 +4112,21 @@ tBTM_EIR_SEARCH_RESULT BTM_HasInquiryEirService( tBTM_INQ_RESULTS *p_results,
+** Returns          None
+**
+*******************************************************************************/
+//extern
```
same here

xiongweichao commented on components/bt/host/bluedroid/stack/include/stack/btm_api.h:
```diff
@@ -4110,6 +4142,21 @@ void BTM_AddEirService( UINT32 *p_eir_uuid, UINT16 uuid16 );
+** Returns          None
+**
+*******************************************************************************/
+//extern
```
same here

xiongweichao commented on components/bt/host/bluedroid/stack/btm/btm_inq.c:
```diff
@@ -2535,9 +2535,125 @@ void BTM_AddEirService( UINT32 *p_eir_uuid, UINT16 uuid16 )
+static bool btm_compare_uuid(tBT_UUID *uuid1, tBT_UUID *uuid2)
+{
+    if (uuid1->len != uuid2->len) {
+        return false;
```
In order to maintain a consistent code style, it is recommended to change it to `FALSE`.

ESP-YTGerd commented:
Hi, @ilutchenko The internal MR in gitlab was created and all the changes will be automatically synchronized to github, once the MR is merged.

ESP-YTGerd commented:
Hi, @ilutchenko 
This PR has been merged and thanks for your contribution again.
